### ISSUE TITLE
Matching request fails due to deleted donor (KeyNotFoundException)

### ISSUE THREAD

**Issue by zabeen** [Anthony-Nolan/Atlas#1055]
[Same problem was observed at match prediction, `PersistSearchResults` stage](#1030): the fix was to supply the required donor info to match prediction via the matching result, to save making the additional call to the donor store (#728).

However, the matching component itself is susceptible to the same "key not found" exception, when a matching donor is deleted from the donor store before its metadata can be retrieved.

Normally this is not an issue, and the request will simply replay until the matching donor db has caught up. However, last night my test search on LIVE-WMDA-ATLAS failed 10 times with this exception because there were a large number of searchable donor updates (over 100K) waiting to be applied to the active db.

The same exception was also observed for other searches run on live earlier in the day:
770e81bc-a8be-4f92-b224-4c022f59b35e
4c3462d6-4b88-4b1c-8641-376195048892
c02b998d-1968-4644-aa3f-1029ddaad3c5
517f4f64-b70c-4f6b-aadc-90c31d0c4f1e
8c65aec7-132c-4ba6-a757-392110432ad5
e02baf9a-e207-48d0-a8ac-6ab6787e7e8b

There were over 155K records in `Donors.PublishableDonorUpdates` with datetime of 5th Sept 2023, 11am.

Running this query, I can see that there are several occurrences where thousands of updates have been created in the space of an hour:
```sql
select FORMAT(CreatedOn, 'yyyy-MM-dd HH') as CreatedOnDay, count(*)
from Donors.PublishableDonorUpdates
group by FORMAT(CreatedOn, 'yyyy-MM-dd HH')
order by CreatedOnDay
```

So this is a real, live scenario that Atlas will face regularly.

**Comment by IgorKupreychik** [comment]
I performed additional tests in UAT environment. First I changed CRON schedule for publishing Donor import updates to run every 12 seconds (*/12 * * * * *), then updated batch size for Matching Donor Management Function to 10000.
Data set is 225000 donors, all are new (i.e. insert data set).
The file with donors was uploaded at 5:18 AM UTC, overall processing finished at 5:52 AM. Looks like everything was processed successfully, there are no new records in alerts topic.

Details:

- PublishSearchableDonorUpdates
  -- max processing time: 29 sec
  -- average processing time: 6 sec
  -- finish: 5:46 AM
- ProcessDifferentialDonorUpdatesForMatchingDbA
  -- max processing time: 22 sec
  -- average processing time: 8-9 sec
  -- finish: 5:49 AM
  -- max number of items waiting in the queue: < 20000
- ProcessDifferentialDonorUpdatesForMatchingDbB
  -- max processing time: 76 sec
  -- average processing time: 50-60 sec, but sometimes it was 10 sec
  -- finish: 5:52 AM
  -- max number of items waiting in the queue: 60000

Results for donor removal dataset (same 225000 donors, that were added before):
Start: 6:05 AM, Finish: 6:37 AM. Looks like everything was processed successfully, there are no new records in alerts topic.
Details:

- PublishSearchableDonorUpdates
  -- max processing time: 27 sec
  -- average processing time: 5 sec
  -- finish: 6:31 AM
- ProcessDifferentialDonorUpdatesForMatchingDbA
  -- max processing time: 12 sec
  -- average processing time: 7-8 sec
  -- finish: 6:36 AM
  -- max number of items waiting in the queue: < 20000
- ProcessDifferentialDonorUpdatesForMatchingDbB
  -- max processing time: 27 sec
  -- average processing time: 20-25 sec, but sometimes it was <10 sec
  -- finish: 6:37 AM
  -- max number of items waiting in the queue: < 20000
